pkg/open/open.go

```diff
@@ -0,0 +1,238 @@
+package open
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/glasskube/glasskube/api/v1alpha1"
+	"github.com/glasskube/glasskube/pkg/client"
+	"github.com/glasskube/glasskube/pkg/future"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+type opener struct {
+	pkgClient  *client.PackageV1Alpha1Client
+	ksClient   kubernetes.Interface
+	restClient rest.Interface
+	stopCh     []chan struct{}
+	readyCh    []chan struct{}
+	stopped    bool
+}
+
+func NewOpener() *opener {
+	return &opener{}
+}
+
+func (o *opener) Open(ctx context.Context, packageName string, entrypointName string) (*openResult, error) {
+	if err := o.initFromContext(ctx); err != nil {
+		return nil, err
+	}
+
+	var pkg v1alpha1.Package
+	if err := o.pkgClient.Packages().Get(ctx, packageName, &pkg); err != nil {
+		return nil, fmt.Errorf("could not get package %v: %w", packageName, err)
+	}
+
+	var packageInfo v1alpha1.PackageInfo
+	if err := o.pkgClient.PackageInfos().Get(ctx, pkg.Spec.PackageInfo.Name, &packageInfo); err != nil {
+		return nil, fmt.Errorf("could not get PackageInfo for package %v: %w", packageName, err)
+	}
+
+	entrypoints := packageInfo.Status.Manifest.Entrypoints
+	if len(entrypoints) < 1 {
+		return nil, fmt.Errorf("package %v has no entrypoint", packageName)
+	}
+
+	if entrypointName != "" {
+		exists := false
+		for _, entrypoint := range entrypoints {
+			if entrypoint.Name == entrypointName {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			return nil, fmt.Errorf("package %v has no entrypoint %v", packageName, entrypointName)
+		}
+	}
+
+	result := openResult{opener: o}
+	var futures []future.Future
+	for _, entrypoint := range entrypoints {
+		if entrypointName == "" || entrypoint.Name == entrypointName {
+			e := entrypoint
+			readyCh := make(chan struct{})
+			stopCh := make(chan struct{})
+			o.readyCh = append(o.readyCh, readyCh)
+			o.stopCh = append(o.stopCh, stopCh)
+			entrypointFuture, err := o.open(ctx, packageInfo.Status.Manifest, e, readyCh, stopCh)
```

Comment:
A port conflict will lead to a silent failure. (e.g. executing `docusaurus start` will listen on port 3000 and `glasskube open cyclops` will also try to create a port forward for this specific port). As of now this is a silent error and the docusaurus website is just opened in the browser. We should exit the open command with an error.

Reply:
That's interesting, because it does that for me…
![image](https://github.com/glasskube/glasskube/assets/16959694/20dd17f2-f52d-4f95-80e3-9e2e626a0e8a)

Reply:
I introduced an additional check for `:[port]` to catch this problem.